### PR TITLE
kata-deploy: fix build for arm64

### PR DESCRIPTION
**Pull request by liyimeng**
Current build script hard written to x86_64.

### CODE REVIEW

**Comment by liyimeng** [comment]
This need a arm64 machine to build natively, not cross compilation tested yet.

**Comment by liyimeng** [comment]
https://github.com/kata-containers/kata-containers/issues/5507

**Comment by liyimeng** [comment]
Also update kata-deploy base images into docker.io/centos:7. which have multi-arch suported. 
Meanwhile,  optimised kata-deploy images size.

File: tools/packaging/kata-deploy/Dockerfile [hunk 1]
```diff
@@ -26,4 +17,6 @@ echo "gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packa
 yum -y install kubectl && \
 yum clean all
 
+RUN mkdir -p ${DESTINATION}
+ADD ${KATA_ARTIFACTS} ${DESTINATION}/
```
Comment (liyimeng) [inline tools/packaging/kata-deploy/Dockerfile:21]: CI tools dose not like this:

```
tools/packaging/kata-deploy/Dockerfile:21 DL3020 error: Use COPY instead of ADD for files and folders
[1410](https://github.com/kata-containers/kata-containers/actions/runs/3345467724/jobs/5566985626#step:11:1411)
ERROR: failed to check Dockerfile 'tools/packaging/kata-deploy/Dockerfile'
```

Not sure why this policy is enforced. Using ADD instead here, docker will automatically extract files into target, hence reduce the size of the images with about 100MB.

File: tools/packaging/kata-deploy/Dockerfile [hunk 2]
```diff
@@ -4,19 +4,10 @@
 
 # Specify alternative base image, e.g. clefos for s390x
 ARG IMAGE
-FROM ${IMAGE:-registry.centos.org/centos}:7
-ARG KATA_ARTIFACTS=./kata-static.tar.xz
+FROM ${IMAGE:-docker.io/centos}:7
```
Comment (fidencio) [inline tools/packaging/kata-deploy/Dockerfile:7]: What's the reason for switching this to docker.io?  registry.centros.org doesn't provide an ARM image?

This should be explained as part of the commit message.

File: tools/packaging/kata-deploy/Dockerfile [hunk 3]
```diff
@@ -4,19 +4,10 @@
...
+ARG KATA_ARTIFACTS=./local-build//kata-static.tar.xz
```
Comment (fidencio) [inline tools/packaging/kata-deploy/Dockerfile:8]: Could you explain why we need this?

File: tools/packaging/static-build/firecracker/build-static-firecracker.sh [hunk 4]
```diff
@@ -37,5 +37,5 @@ git fetch
 git checkout ${firecracker_version}
 sudo ./tools/devtool --unattended build --release
 
-ln -sf ./build/cargo_target/x86_64-unknown-linux-musl/release/firecracker ./firecracker-static
-ln -sf ./build/cargo_target/x86_64-unknown-linux-musl/release/jailer ./jailer-static
+ln -sf ./build/cargo_target/$(uname -m)-unknown-linux-musl/release/firecracker ./firecracker-static
+ln -sf ./build/cargo_target/$(uname -m)-unknown-linux-musl/release/jailer ./jailer-static
```
Comment (fidencio) [inline tools/packaging/static-build/firecracker/build-static-firecracker.sh:41]: Thiese chamges would be better to come as a different commit, still par of this very same PR.

**Review by fidencio — COMMENTED** [review]
I've left a few questions on the PR.

File: tools/packaging/kata-deploy/Dockerfile [hunk 5]
```diff
@@ -4,19 +4,10 @@
...
 ARG DESTINATION=/opt/kata-artifacts
 
-COPY ${KATA_ARTIFACTS} ${WORKDIR}
-
-RUN \
-yum -y update && \
```
Comment (fidencio) [inline tools/packaging/kata-deploy/Dockerfile:14]: Why did you remove this?

We should always be looking for package fixes, which would come from a yum update.

File: tools/packaging/kata-deploy/Dockerfile (hunk 3 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:8]: @fidencio  I test it. registry.centos.org/centos is not a manifest tag, i.e. it only provides x86_64 under that tag. docker.io provide all archs, hence go with docker.io

File: tools/packaging/kata-deploy/Dockerfile (hunk 2 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:7]: @fidencio I test it. registry.centos.org/centos is not a manifest tag, i.e. it only provides x86_64 under that tag. docker.io provide all archs, hence go with docker.io

File: tools/packaging/kata-deploy/Dockerfile (hunk 3 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:8]: Make it east to for developer. With this change, developer just run docker build after all staic files built, no need an extra step to move the static file into target place. Less work, less problem.

File: tools/packaging/kata-deploy/Dockerfile (hunk 5 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:14]: After replacing COPY with ADD, this become necessary. Remove it to save image size. 
See reference https://github.com/kata-containers/kata-containers/pull/5534#discussion_r1009103543

File: tools/packaging/kata-deploy/Dockerfile (hunk 2 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:7]: addressed with new commit message.

File: tools/packaging/static-build/firecracker/build-static-firecracker.sh (hunk 4 above)
Comment (liyimeng) [reply tools/packaging/static-build/firecracker/build-static-firecracker.sh:41]: It is kind of minor change, I have no opinion. If it is a better kata coding style, not a problem for me.

Comment (liyimeng) [reply tools/packaging/static-build/firecracker/build-static-firecracker.sh:41]: addressed in new update as suggested @fidencio

File: tools/packaging/kata-deploy/Dockerfile (hunk 5 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:14]: If we need fixes, like emergency security fixes, we should update the base image instead of using unpredictable yum update.  In a container world, immutable is important feature. Just my two cents though.

**Comment by liyimeng** [comment]
@fidencio any further comments?

File: tools/packaging/kata-deploy/Dockerfile (hunk 5 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/Dockerfile:14]: And we're working on a different way to address this.

There's been a whole bunch of work in the CCv0 branch, which will end up merged also in the `main` branch, to cache the images used and provide those as part of the release process, so users can easily reproduce the build.

For now, please, leave it as it's.

File: tools/packaging/kata-deploy/Dockerfile (hunk 1 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/Dockerfile:21]: This seems to be an static check warning that we could ignore.

**Comment by fidencio** [comment]
In general this work looks very good to me.
@wainersm, what's the best way to ignore the hadolint error in a specific file?

File: tools/packaging/kata-deploy/Dockerfile (hunk 5 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/Dockerfile:14]: @liyimeng, any further updates here?

Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:14]: @fidencio Not sure if I am getting you. Do you want to get back line 11-14, which I removed in this commit?

**Comment by liyimeng** [comment]
What to do to get this move on?

File: tools/packaging/kata-deploy/Dockerfile (hunk 5 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/Dockerfile:14]: Yes.

Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:14]: hmm, I still think replace these with line 20 21 as below is more elegant way to do thing. If you insist, I could put pack line line 13 and 14. 

Honestly, If we do care security, this image should run rootless and something other more proper solution. This update dose not add much value here.

File: tools/packaging/kata-deploy/Dockerfile (hunk 1 above)
Comment (liyimeng) [reply tools/packaging/kata-deploy/Dockerfile:21]: @wainersm  may you please help us to get this move on.

**Review by liyimeng — COMMENTED** [review]
@wainersm Please kindly help us out.

**Comment by fidencio** [comment]
This one has been solved in a different way in the past releases, and we're closing it for now.
Please, feel free to open a new PR in case something was not addressed.
